jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -495,16 +510,33 @@ public MethodHandle getBoundMethodHandle() {
 			MethodType nativeMethodType = methodType(Object.class, Addressable.class, SegmentAllocator.class, long[].class);
 			/*[ENDIF] JAVA_SPEC_VERSION >= 21 */
 
-			MethodHandle boundHandle = lookup.bind(this, "runNativeMethod", nativeMethodType);
+			boundMH = lookup.bind(this, "runNativeMethod", nativeMethodType);
 
 			/* Replace the original handle with the specified types of the C function. */
-			boundHandle = permuteMH(boundHandle, funcMethodType);
-			return boundHandle;
+			boundMH = permuteMH(boundMH, funcMethodType);
+
+			/*[IF JAVA_SPEC_VERSION >= 22]*/
+			return isFfiProtoOn ? getNativeMHWithInvokeCache(boundMH) : boundMH;
+			/*[ELSE] JAVA_SPEC_VERSION >= 22 */
+			return boundMH;
+			/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
 		} catch (ReflectiveOperationException e) {
 			throw new InternalError(e);
 		}
 	}
 
+	/*[IF JAVA_SPEC_VERSION >= 22]*/
+	/* Generate the native MethodHandle with the resolved bound handle. */
+	private MethodHandle getNativeMHWithInvokeCache(MethodHandle handle) {
+		nativeMH = SharedSecrets.getJavaLangInvokeAccess().nativeMethodHandle(handle);
+		/* The cache array declared in the NativeMethodHandle (OpenJDK) stores MemberName
+		 * and appendix resolved from the bounded handle.
+		 */
+		resolveBoundHandle();
```

Comment:
does this have to be done in native code?

Reply:
Yes as there is no way around to deal with the tricky situation given we need to store the cache array for the bound MH in the native MH at https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/a9c216c088a436d36d36181f3300abd5046c4147/src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java#L58, in which case the `memberName` for the bound MH will be extracted from this cache array (stored in the native MH) in `LinkToNative` in the interpreter.

Reply:
ah okay, thanks for explaining that.